File: src/main/java/com/cobblegyms/config/CobbleGymsConfig.java
package com.cobblegyms.config;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import net.fabricmc.loader.api.FabricLoader;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

public class CobbleGymsConfig {
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private static CobbleGymsConfig instance;

    // Season settings
    public int seasonDurationDays = 30;
    public int minBattlesForReward = 25;
    public double minWinrateForReward = 0.5;

    // Challenge settings
    public long challengeCooldownSeconds = 86400L; // 24 hours
    public int tempBanMaxHours = 72;

    // Discord settings
    public String discordWebhookUrl = "";
    public String discordBotToken = "";
    public Map<String, String> leaderDiscordChannels = new HashMap<>();

    // Gym locations per type
    public Map<String, GymLocation> gymLocations = new HashMap<>();

    // Reward settings
    public String gymLeaderWeeklyRewardCommand = "give %player% minecraft:diamond 5";
    public String e4WeeklyRewardCommand = "give %player% minecraft:diamond_block 1";
    public String championWeeklyRewardCommand = "give %player% minecraft:netherite_ingot 1";
    public String seasonEndRewardCommand = "give %player% minecraft:gold_ingot %amount%";

    public static class GymLocation {
        public double x, y, z;
        public String world = "minecraft:overworld";

        public GymLocation() {}
        public GymLocation(double x, double y, double z, String world) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.world = world;
        }
    }

    private CobbleGymsConfig() {}

    public static CobbleGymsConfig getInstance() {
        if (instance == null) {
            instance = load();
        }
        return instance;
    }

    public static CobbleGymsConfig load() {
        Path configDir = FabricLoader.getInstance().getConfigDir().resolve("cobblegyms");
        Path configFile = configDir.resolve("config.json");

        try {
            if (!Files.exists(configDir)) {
                Files.createDirectories(configDir);
            }

            if (!Files.exists(configFile)) {
                CobbleGymsConfig defaults = new CobbleGymsConfig();
                defaults.save();
                instance = defaults;
                return defaults;
            }

            try (Reader reader = Files.newBufferedReader(configFile)) {
                CobbleGymsConfig config = GSON.fromJson(reader, CobbleGymsConfig.class);
                if (config == null) config = new CobbleGymsConfig();
                instance = config;
                return config;
            }
        } catch (IOException e) {
            System.err.println("[CobbleGyms] Failed to load config: " + e.getMessage());
            CobbleGymsConfig defaults = new CobbleGymsConfig();
            instance = defaults;
            return defaults;
        }
    }

    public void save() {
        Path configDir = FabricLoader.getInstance().getConfigDir().resolve("cobblegyms");
        Path configFile = configDir.resolve("config.json");

        try {
            if (!Files.exists(configDir)) {
                Files.createDirectories(configDir);
            }
            try (Writer writer = Files.newBufferedWriter(configFile)) {
                GSON.toJson(this, writer);
            }
        } catch (IOException e) {
            System.err.println("[CobbleGyms] Failed to save config: " + e.getMessage());
        }
    }

    public long getSeasonDurationMillis() {
        return (long) seasonDurationDays * 24 * 60 * 60 * 1000;
    }

    public long getChallengeCooldownMillis() {
        return challengeCooldownSeconds * 1000;
    }
}

File: src/main/kotlin/com/cobblegyms/config/CobbleGymsConfig.kt
package com.cobblegyms.config

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileReader
import java.io.FileWriter

/**
 * Main configuration for CobbleGyms loaded from cobblegyms.json.
 */
data class CobbleGymsConfig(
    val database: DatabaseConfig = DatabaseConfig(),
    val season: SeasonConfig = SeasonConfig(),
    val battle: BattleConfig = BattleConfig(),
    val rewards: RewardsConfig = RewardsConfig(),
    val smogon: SmogonConfig = SmogonConfig(),
    val discord: DiscordConfig = DiscordConfig(),
    val gui: GuiConfig = GuiConfig()
) {
    data class DatabaseConfig(
        val filePath: String = "config/cobblegyms/cobblegyms.db"
    )

    data class SeasonConfig(
        val durationDays: Int = 30,
        val autoReset: Boolean = true
    )

    data class BattleConfig(
        val queueTimeoutSeconds: Long = 300,
        val maxQueueSize: Int = 50,
        val allowSpectators: Boolean = true
    )

    data class RewardsConfig(
        val gymBadgeReward: Int = 1000,
        val e4WinReward: Int = 5000,
        val championReward: Int = 10000,
        val cooldownHours: Int = 24
    )

    data class SmogonConfig(
        val format: String = "gen9ou",
        val banListEnabled: Boolean = true,
        val bannedPokemon: List<String> = listOf(
            "Koraidon", "Miraidon", "Zacian", "Zamazenta", "Eternatus",
            "Kyogre", "Groudon", "Rayquaza", "Dialga", "Palkia", "Giratina",
            "Reshiram", "Zekrom", "Kyurem-White", "Kyurem-Black",
            "Xerneas", "Yveltal", "Zygarde-Complete",
            "Cosmog", "Cosmoem", "Solgaleo", "Lunala", "Necrozma-Dawn",
            "Necrozma-Dusk", "Calyrex-Shadow", "Calyrex-Ice"
        ),
        val bannedMoves: List<String> = listOf(
            "Baton Pass"
        ),
        val bannedAbilities: List<String> = listOf(
            "Arena Trap", "Shadow Tag", "Moody", "Power Construct"
        ),
        val teamSize: Int = 6,
        val maxLevel: Int = 100
    )

    data class DiscordConfig(
        val enabled: Boolean = false,
        val botToken: String = "",
        val channelId: String = "",
        val notifyGymWins: Boolean = true,
        val notifyE4Wins: Boolean = true,
        val notifyChampion: Boolean = true,
        val notifySeasonEnd: Boolean = true
    )

    data class GuiConfig(
        val useGui: Boolean = true,
        val mainMenuTitle: String = "§6✦ CobbleGyms ✦",
        val gymMenuTitle: String = "§e⚔ Gym Leaders",
        val e4MenuTitle: String = "§5★ Elite Four",
        val challengeMenuTitle: String = "§c⚡ Challenge Queue"
    )

    companion object {
        private val logger = LoggerFactory.getLogger("CobbleGyms/Config")
        private val gson: Gson = GsonBuilder().setPrettyPrinting().create()
        private const val CONFIG_FILE = "config/cobblegyms/config.json"

        private var instance: CobbleGymsConfig = CobbleGymsConfig()

        fun load(): CobbleGymsConfig {
            val file = File(CONFIG_FILE)
            if (!file.exists()) {
                logger.info("No config found, creating default config at $CONFIG_FILE")
                save(CobbleGymsConfig())
                return CobbleGymsConfig()
            }
            return try {
                FileReader(file).use { reader ->
                    gson.fromJson(reader, CobbleGymsConfig::class.java).also {
                        instance = it
                        logger.info("Configuration loaded from $CONFIG_FILE")
                    }
                }
            } catch (e: Exception) {
                logger.error("Failed to load config, using defaults: ${e.message}")
                CobbleGymsConfig()
            }
        }

        fun save(config: CobbleGymsConfig = instance) {
            val file = File(CONFIG_FILE)
            file.parentFile?.mkdirs()
            try {
                FileWriter(file).use { writer ->
                    gson.toJson(config, writer)
                }
                instance = config
                logger.info("Configuration saved to $CONFIG_FILE")
            } catch (e: Exception) {
                logger.error("Failed to save config: ${e.message}")
            }
        }

        fun get(): CobbleGymsConfig = instance
    }
}
